UART Driver/Core/Src/main.c

```diff
@@ -0,0 +1,307 @@
+/* USER CODE BEGIN Header */
+/**
+  ******************************************************************************
+  * @file           : main.c
+  * @brief          : Main program body
+  ******************************************************************************
+  * @attention
+  *
+  * Copyright (c) 2022 STMicroelectronics.
+  * All rights reserved.
+  *
+  * This software is licensed under terms that can be found in the LICENSE file
+  * in the root directory of this software component.
+  * If no LICENSE file comes with this software, it is provided AS-IS.
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+UART_HandleTypeDef huart2;
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_USART2_UART_Init(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+//Creates the fijo and foji struct
+//The reason there are 2 different gpsCoordinates structs is to minimize the size of the fijo struct
+struct gpsCoordinatesFOJI{
+	double lattitude;
+	double longtitude;
+	double altitude;
+};
+
+struct gpsCoordinatesFIJO{
+	double lattitude;
+	double longtitude;
+};
+
+struct eulerAnglesOfPlane{
+	float yaw;
+	float pitch;
+	float roll;
+};
+
+//Should be 37 bytes but because of padding its really 48 bytes
```

Comment:
I'm OOTL, could you explain the padding situation?

Reply:
Sure, here's the gist of it: in this case, a 64bit processor can access 8 bytes at a time, so in this example, when reading the foji struct my first byte would be my start, and the next data type in my struct is a double. But the processor only has access to 7 bytes left since the start already took one up. Since a double is 8 bytes it cant fit. This means 7 bytes would be left empty. Then all the doubles would be accessed smoothly without empty bytes because they are all 8 bytes. Then when reading the floats the first 2 floats would be read without empty bytes as they sum up to 8. The last float would be 4 bytes which means 4 empty left. This means a total of 4 + 7 = 11 bytes were left empty. That's why the size is not 37 but 11+37 = 48. Lmk if that helps.